codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/EndpointHostPrefixMiddleware.java

```diff
@@ -0,0 +1,204 @@
+            w.openBlock("if !ok {", "}", () -> {
+                writer.write("return out, metadata, fmt.Errorf(\"unknown transport type %T\", in.Request)");
+            }).write("");
+
+            if (pattern.getLabels().isEmpty()) {
+                w.write("request.HostPrefix = $S", pattern.toString());
+
+            } else {
+                // If the pattern has labels, we need to build up the host prefix using a string builder.
+                writer.addUseImports(SmithyGoDependency.STRINGS);
+
+                w.write("var prefix strings.Builder");
+                StructureShape input = ProtocolUtils.expectInput(model, operation);
+                for (SmithyPattern.Segment segment : pattern.getSegments()) {
+                    if (!segment.isLabel()) {
+                        w.write("prefix.WriteString($S)", segment.toString());
+                    } else {
+                        MemberShape member = input.getMember(segment.getContent()).get();
+                        String memberName = symbolProvider.toMemberName(member);
+                        String memberReference = "input." + memberName;
+
+                        // Theoretically this should never be nil or empty by this point unless validation has
+                        // been disabled.
+                        w.write("if $L == nil || len(*$L) == 0 {", memberReference).indent();
+                        w.write("return out, metadata, &smithy.SerializationError{Err: "
+                                        + "fmt.Errorf(\"$L forms part of the endpoint host and so may not be nil\")}",
+                                memberName);
+                        w.dedent().write("} else if !smithyhttp.ValidHostLabel(*$L) {", memberReference).indent();
+                        w.write("return out, metadata, &smithy.SerializationError{Err: "
+                                + "fmt.Errorf(\"$L forms part of the endpoint host and so must match "
+                                + "\\\"[a-zA-Z0-9-]{1,63}\\\""
+                                + ", but was \\\"%s\\\"\", *$L)}", memberName, memberReference);
+                        w.dedent().openBlock("} else {", "}", () -> {
+                            w.write("prefix.WriteString(*$L)", memberReference);
+                        });
+                    }
+                }
+
+                w.write("request.HostPrefix = prefix.String()");
+            }
+            w.write("");
+
+            w.write("return next.$L(ctx, in)", generator.getHandleMethodName());
+        });
+    }
+
+    /**
+     * Gets a list of the operations decorated with the EndpointTrait.
+     *
+     * @param model   Model used for generation.
+     * @param service Service for getting list of operations.
+     * @return list of operations decorated with the EndpointTrait.
+     */
+    public static List<OperationShape> getOperationsWithEndpointPrefix(Model model, ServiceShape service) {
```

Comment:
This can be `private` method

Reply:
This was made public so that tests could be generated in the SDK's codegen module.